### ISSUE TITLE
ArgumentNullException when sending data to kinesis in ASP.NET Core 8 Web API

### ISSUE THREAD

**Issue by neox2811** [aws/aws-sdk-net#3110]
After upgrading my project to .NET 8.0, the call of the function PutRecordsAsync() from AWSSDK.Kinesis (3.7.300.7) results in the following error:

```
System.Private.CoreLib Value cannot be null. (Parameter 'path1') System.Collections.ListDictionaryInternal     at System.ArgumentNullException.Throw(String paramName)
   at System.IO.Path.Combine(String path1, String path2)
   at Amazon.Internal.RegionEndpointProviderV3.GetEndpointJsonSourceStream()
   at Amazon.Internal.RegionEndpointProviderV3..ctor()
   at Amazon.RegionEndpoint.get_RegionEndpointProvider()
   at Amazon.RegionEndpoint.get_InternedRegionEndpoint()
   at Amazon.RegionEndpoint.GetEndpointForService(String serviceName, GetEndpointForServiceOptions options)
   at Amazon.Runtime.Internal.Auth.AWS4Signer.DetermineSigningRegion(IClientConfig clientConfig, String serviceName, RegionEndpoint alternateEndpoint, IRequest request)
   at Amazon.Runtime.Internal.Auth.AWS4Signer.SignRequest(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, String awsAccessKeyId, String awsSecretAccessKey)
   at Amazon.Runtime.Internal.Auth.AWS4Signer.Sign(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, String awsAccessKeyId, String awsSecretAccessKey)
   at Amazon.Runtime.Internal.Auth.AWS4Signer.Sign(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, ImmutableCredentials credentials)
   at Amazon.Runtime.Internal.Auth.AbstractAWSSigner.SignAsync(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, ImmutableCredentials credentials, CancellationToken token)
   at Amazon.Runtime.Internal.Signer.SignRequestAsync(IRequestContext requestContext)
   at Amazon.Runtime.Internal.Signer.PreInvokeAsync(IExecutionContext executionContext)
   at Amazon.Runtime.Internal.Signer.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CredentialsRetriever.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.MetricsHandler.InvokeAsync[T](IExecutionContext executionContext)
   at <mystacktrace>.SendeDatenAnKinesis(String streamName, Int32 bufferSize)
```

Code:

`var putRecordsResponse = await _kinesisClient.PutRecordsAsync(putRecordsRequest);`

What I already checked:

putRecordsRequest is not null and filled with valid values
_kinesisClient is correctly initialized with credentials & configuration
With .NET 7 everything (same code, no changes) worked as expected. Does the AWS SDK support .NET 8.0 at the moment? As stated on nuget, .NET 8.0 should be supported.

https://www.nuget.org/packages/AWSSDK.Kinesis -> .NET 8.0 .NET Core 3.1 .NET Standard 2.0 .NET Framework 3.5

_Originally posted by @neox2811 in https://github.com/aws/aws-sdk-net/discussions/3109_

Also see https://stackoverflow.com/questions/77530522/argumentnullexception-when-sending-data-to-kinesis-in-asp-net-core-8-web-api/

**Comment by dscpinheiro** [comment]
I know this was converted from the discussion, but can you add more detailed reproduction steps? Here's what we'd need (from the `Bug Report` template):
```
Provide a self-contained, concise snippet of code that can be used to reproduce the issue.
For more complex issues provide a repo with the smallest sample that reproduces the bug.

Avoid including business logic or unrelated code, it makes diagnosis more difficult.
The code sample should be an SSCCE. See http://sscce.org/ for details. In short, please provide a code sample that we can copy/paste, run and reproduce.
```

The environment details (e.g. Operating System and version) will also be helpful.

**Comment by ashishdhingra** [comment]
@neox2811 Good afternoon. Somehow, I'm unable to reproduce using the below code (used `AWSSDK.Kinesis` version `3.7.300.7`):
```C#
using Amazon.Kinesis;
using Amazon.Kinesis.Model;
using System.Net.Http.Headers;
using System.Text;

using (var _kinesisClient = new AmazonKinesisClient())
{
    var streamName = "testdatastream";
    var recordslist = new List<PutRecordsRequestEntry>();

    for (int j = 0; j < 10; ++j)
    {
        PutRecordsRequestEntry requestRecord = new PutRecordsRequestEntry();
        requestRecord.Data = new MemoryStream(Encoding.UTF8.GetBytes("testData-" + j));
        requestRecord.PartitionKey = "partitionKey-" + j;
        recordslist.Add(requestRecord);
    }

    PutRecordsRequest putRecordsRequest = new PutRecordsRequest()
    {
        StreamName = streamName,
        Records = recordslist
    };
    var putRecordsResponse = await _kinesisClient.PutRecordsAsync(putRecordsRequest);
}
```
Here is what my `.csproj` file looks like:
```XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <PublishAot>true</PublishAot>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="AWSSDK.Kinesis" Version="3.7.300.7" />
  </ItemGroup>

</Project>
```

Could you please share the version of `AWSSDK.Kinesis` package in your project?

Thanks,
Ashish

**Comment by neox2811** [comment]
Hi,

version of `AWSSDK.Kinesis` = **3.7.301**

I cannot share all the code, but this are the parts that are causing the error with .NET 8:

Service Extension Method, for adding the KinesisConnector in Startup (`services.AddKinesisConnector();`)

**KinesisConnectorServiceExtensions.cs**

```
public static void AddKinesisConnector(this IServiceCollection services)
    {
        var awsConfig = new AmazonKinesisConfig
        {
            RegionEndpoint = Amazon.RegionEndpoint.EUCentral1
        };

        var accessKey = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY");
        var secretKey = Environment.GetEnvironmentVariable("AWS_SECRET_KEY");

        if (accessKey == null || secretKey == null)
        {
            services.AddSingleton<IKinesisConnector, KinesisFakeConnector>();
        }
        else
        {
            var awsCredentials = new BasicAWSCredentials(accessKey, secretKey);
            var client = new AmazonKinesisClient(awsCredentials, awsConfig);

            services.AddSingleton(client);
            services.AddSingleton<IKinesisConnector, KinesisConnector>();
        }
    }
```

**KinesisConnector.cs**

`_kinesisClient` as configured above is injected via DI

```
private readonly ConcurrentDictionary<string, BlockingCollection<PutRecordsRequestEntry>> _buffer = new();

public async Task SendRecord(string streamName, IKinesisLogInput logInput, int bufferSize = 10)
    {
        if (!_buffer.ContainsKey(streamName))
        {
            _buffer.GetOrAdd(streamName, new BlockingCollection<PutRecordsRequestEntry>());
        }

        var jsonData = JsonConvert.SerializeObject(logInput);

        var requestRecord = new PutRecordsRequestEntry
        {
            Data = new MemoryStream(Encoding.UTF8.GetBytes(jsonData)),
            PartitionKey = Guid.NewGuid().ToString()
        };

        _buffer[streamName].Add(requestRecord);

        await SendToKinesis(streamName, bufferSize);
    }

private async Task SendToKinesis(string streamName, int bufferSize = 0)
    {
        PutRecordsRequest putRecordsRequest = null;
        lock (_lock)
        {
            if (_buffer[streamName].Count > bufferSize)
            {
                putRecordsRequest = new PutRecordsRequest
                {
                    StreamName = streamName,
                    Records = _buffer[streamName].ToList(),
                };
                _buffer[streamName] = new BlockingCollection<PutRecordsRequestEntry>();
            }
        }

        if (putRecordsRequest != null)
        {
            try
            {
// This line causes the exception                
var putRecordsResponse = await _kinesisClient.PutRecordsAsync(putRecordsRequest);

                if (putRecordsResponse.FailedRecordCount > 0)
                {
                    foreach (var failedRecord in putRecordsResponse.Records)
                    {
                        _logger.LogError(
                            $"Error for {streamName}: {failedRecord.ErrorCode} {failedRecord.ErrorMessage}");
                    }
                }
            }
            catch (Exception exception)
            {
                _logger.LogError(exception, exception.Message);
            }
        }
    }
```

**Comment by aaronolds** [comment]
I am not using Kinensis, but I am getting the same exact error in DynamoDb Client (Include="AWSSDK.DynamoDBv2" Version="3.7.300.7").  However, I am unable to reproduce this issue on my local development machines (M1 Mac Pro and Windows 11).  I am only seeing this issue when I run my services in Linux docker images on EKS. 

Message":"Value cannot be null. (Parameter 'path1')

at System.ArgumentNullException.Throw(String paramName)
   at System.IO.Path.Combine(String path1, String path2)
   at Amazon.Internal.RegionEndpointProviderV3.GetEndpointJsonSourceStream()
   at Amazon.Internal.RegionEndpointProviderV3..ctor()
   at Amazon.RegionEndpoint.get_RegionEndpointProvider()
   at Amazon.RegionEndpoint.get_InternedRegionEndpoint()
   at Amazon.RegionEndpoint.GetEndpointForService(String serviceName, GetEndpointForServiceOptions options)
   at Amazon.Runtime.Internal.Auth.AWS4Signer.DetermineSigningRegion(IClientConfig clientConfig, String serviceName, RegionEndpoint alternateEndpoint, IRequest request)
   at Amazon.Runtime.Internal.Auth.AWS4Signer.SignRequest(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, String awsAccessKeyId, String awsSecretAccessKey)
   at Amazon.Runtime.Internal.Auth.AWS4Signer.Sign(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, String awsAccessKeyId, String awsSecretAccessKey)
   at Amazon.Runtime.Internal.Auth.AWS4Signer.Sign(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, ImmutableCredentials credentials)
   at Amazon.Runtime.Internal.Auth.AbstractAWSSigner.SignAsync(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, ImmutableCredentials credentials, CancellationToken token)
   at Amazon.Runtime.Internal.Signer.SignRequestAsync(IRequestContext requestContext)
   at Amazon.Runtime.Internal.Signer.PreInvokeAsync(IExecutionContext executionContext)
   at Amazon.Runtime.Internal.Signer.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CredentialsRetriever.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.MetricsHandler.InvokeAsync[T](IExecutionContext executionContext)
     at <mystacktrace>.SaveStateAsync[TValue](String storeName, String key, TValue value, IReadOnlyDictionary`2 metadata, CancellationToken cancellationToken)

**Comment by dscpinheiro** [comment]
Thanks for the reproduction steps, it looks like it's only happening on Linux (I did not see the error on Windows either). I do see some breaking changes in .NET 8 (https://learn.microsoft.com/en-us/dotnet/core/compatibility/8.0#core-net-libraries) that we might not have accounted for.

@aaronolds I'll close the other issue and keep the investigation here, it's the same problem for both service clients.

**Comment by dscpinheiro** [comment]
If possible, could you share the `Dockerfile` / `.csproj` files you're using? 

I've tried creating a web application (using both the `ASP.NET Core Web API` and `ASP.NET Core Web API - native AOT` templates from Visual Studio) and deployed it to ECS (on Fargate - using `dotnet aws deploy`) but did not see the error.

Here's the files I used:
```Dockerfile
FROM mcr.microsoft.com/dotnet/aspnet:8.0 AS base
USER app
WORKDIR /app
EXPOSE 8080

FROM mcr.microsoft.com/dotnet/sdk:8.0 AS build
# Install clang/zlib1g-dev dependencies for publishing to native
RUN apt-get update \
    && apt-get install -y --no-install-recommends \
    clang zlib1g-dev
ARG BUILD_CONFIGURATION=Release
WORKDIR /src
COPY ["WebApplication2.csproj", "."]
RUN dotnet restore "./WebApplication2.csproj"
COPY . .
WORKDIR "/src/."
RUN dotnet build "./WebApplication2.csproj" -c $BUILD_CONFIGURATION -o /app/build

FROM build AS publish
ARG BUILD_CONFIGURATION=Release
RUN dotnet publish "./WebApplication2.csproj" -c $BUILD_CONFIGURATION -o /app/publish /p:UseAppHost=true

FROM mcr.microsoft.com/dotnet/runtime-deps:8.0 AS final
WORKDIR /app
EXPOSE 8080
COPY --from=publish /app/publish .
ENTRYPOINT ["./WebApplication2"]
```

```csproj
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
    <InvariantGlobalization>true</InvariantGlobalization>
    <PublishAot>true</PublishAot>
    <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
    <DockerfileContext>.</DockerfileContext>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="AWSSDK.Kinesis" Version="3.7.301" />
    <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="[1.19.6-Preview.1, 1.19.6]" />
  </ItemGroup>

</Project>
```

Default route:
```csharp
app.MapGet("/test", async () =>
{
    var kinesisClient = new AmazonKinesisClient(new AmazonKinesisConfig
    {
        RegionEndpoint = Amazon.RegionEndpoint.USWest2
    });

    var randomId = $"minimal-api-{Guid.NewGuid()}";
    using var streamData = new MemoryStream(Encoding.UTF8.GetBytes(randomId));
    var putResponse = await kinesisClient.PutRecordsAsync(new PutRecordsRequest
    {
        Records =
        [
            new PutRecordsRequestEntry
            {
                Data = streamData,
                PartitionKey = randomId,
            }
        ],
        StreamARN = "test-stream-arn",
    });

    return Results.Ok(putResponse.FailedRecordCount);
});
```

**Comment by neox2811** [comment]
Hi @aaronolds,

Have you taken note of the following comment https://stackoverflow.com/a/77531823/6405244 ?

My Dockerfile:
```
FROM mcr.microsoft.com/dotnet/aspnet:8.0
COPY ./publish/ .
ENTRYPOINT ["./MyApp"]
```

I build the app in gitlab ci

```
dotnet publish src/MyApp/ -c Release -o publish
docker build --pull -t $LAN_REGISTRY_IMAGE$DEPLOY_IMAGE .
docker push $LAN_REGISTRY_IMAGE$DEPLOY_IMAGE
```

Then the docker image is pushed to our internal container registry and then run on nomad.

**Comment by aaronolds** [comment]
[neox2811](https://github.com/neox2811),

Thank you for the link to your post on StackOverflow and the comments.   

So it looks like we are both building our applications outside of the docker image and then copying the published assemblies into the image.  Hopefully this helps.

My setup

DockerFile:
```
FROM mcr.microsoft.com/dotnet/aspnet:8.0
COPY "/bin/publish" .
ENTRYPOINT ["dotnet", "myservice.dll"]
```

bitbucket-pipeline.yaml:
```
~~~~~~~~
            # Add nuget sources
            - dotnet nuget add source "https://abc-1234.d.codeartifact.us-east-1.amazonaws.com/nuget/tl/v3/index.json" -n "abc/tl" -u "aws" -p "${CODEARTIFACT_AUTH_TOKEN}" --store-password-in-clear-text
            - dotnet nuget add source "https://abc-1234.d.codeartifact.us-east-1.amazonaws.com/nuget/tc/v3/index.json" -n "abc/tc" -u "aws" -p "${CODEARTIFACT_AUTH_TOKEN}" --store-password-in-clear-text

            # change directory to the project root
            - cd ${PROJECT_DIR}

            # publish
            - dotnet restore
            - dotnet publish --no-restore -c ${CONFIGURATION} -o ${RELEASES_DIR}
            # docker build

            - echo "Building docker image $IMAGE_NAME ..."
            - docker build -f $DOCKER_FILE_LOCATION . -t $IMAGE_NAME

            - echo "Pushing docker image to AWS ECR $IMAGE_NAME ..."
            - aws ecr get-login-password --region ${AWS_DEFAULT_REGION} | docker login --username AWS --password-stdin 1234.dkr.ecr.us-east-1.amazonaws.com
            - docker push $IMAGE_NAME
~~~~~~
```

My DynamoDb Connector csproj:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="AWSSDK.DynamoDBv2" Version="3.7.300.7" />
  </ItemGroup>

</Project>
```

**Comment by dscpinheiro** [comment]
The fix for this issue is available now in the latest package versions (`3.7.300.9` for [DynamoDB](https://www.nuget.org/packages/AWSSDK.DynamoDBv2/3.7.300.9) and `3.7.301.2` for [Kinesis](https://www.nuget.org/packages/AWSSDK.Kinesis/3.7.301.2)).
